### PR TITLE
feat(render.data_frame): Add more methods to `render.data_frame()`

### LINKED ISSUE #1116: Programatically change data frame selection
We need methods to clear and modify data frame row selection from a `reactive.effect`. Probably this should have two methods to set the selected rows and clear all selected rows.

### PR DESCRIPTION
**Pull request by schloerke**
Related #1119 

* Fixes followup to #1120 `mydf.input_selected_rows()`
* Fixes #1116 ; `mydf.update_selected_rows()`
* Adds feature of getting selected data from dataframe renderer. ; `mydf.data_selected_rows()`
* Adds feature of getting data from dataframe renderer. ; `mydf.data()`

### CODE REVIEW

File: shiny/render/_dataframe.py [hunk 1]
```diff
@@ -247,9 +272,12 @@ class data_frame(Renderer[DataFrameResult]):
...
     When using the row selection feature, you can access the selected rows by using the
-    `<data_frame_renderer>.input_selected_rows()` method, where `<data_frame_renderer>` is the render function name that corresponds with the `id=` used in :func:`~shiny.ui.outout_data_frame`. Internally, this method retrieves the selected row value from session's `input.<id>_selected_rows()` value. The value returned will be `None` if no rows
-    are selected, or a tuple of integers representing the indices of the selected rows.
-    To filter a pandas data frame down to the selected rows, use
+    `<data_frame_renderer>.input_selected_rows()` method, where `<data_frame_renderer>`
+    is the render function name that corresponds with the `id=` used in
+    :func:`~shiny.ui.outout_data_frame`. Internally, this method retrieves the selected
+    row value from session's `input.<id>_selected_rows()` value. The value returned will
+    be `None` if no rows are selected, or a tuple of integers representing the indices
+    of the selected rows. To filter a pandas data frame down to the selected rows, use
```
Comment (wch) [inline shiny/render/_dataframe.py:280]: Update with information that a zero-row selection returns an empty tuple; this only returns `None` if row selection is disabled entirely.

File: shiny/render/_dataframe.py [hunk 2]
```diff
@@ -266,27 +294,239 @@ class data_frame(Renderer[DataFrameResult]):
...
-        return active_session.input[self.output_id + "_selected_rows"]()
+        This method is a wrapper around `input.<id>_selected_rows()`, where `<id>` is
+        the `id` of the data frame output. This method returns the selected rows and
+        will cause reactive updates as the selected rows change.
+
+        Returns
+        -------
+        :
+            * `None` if the row selection mode is None
+            * `tuple[int]`a tuple of integers representing the indices of the selected rows
```
Comment (wch) [inline shiny/render/_dataframe.py:499]:
```suggestion
            * `tuple[int]` representing the indices of the selected rows
```

File: shiny/session/_session.py [hunk 3]
```diff
@@ -386,15 +389,25 @@ def _is_hidden(self, name: str) -> bool:
...
             return
 
         try:
             # TODO: handle `blobs`
             value: object = await func(*message["args"])
         except Exception as e:
-            self._send_error_response("Error: " + str(e))
+            print("\n\n")
+            print(message, message["tag"])
+
```
Comment (wch) [inline shiny/session/_session.py:409]: Remove

File: shiny/session/_session.py [hunk 4]
```diff
@@ -404,6 +417,43 @@ async def _send_response(self, message: ClientMessageOther, value: object) -> No
 
     # This is called during __init__.
     def _create_message_handlers(self) -> dict[str, Callable[..., Awaitable[object]]]:
+        # TODO-future; Make sure these methods work within MockSession
+
+        async def outputRPC(
+            outputId: str,
+            handler: str,
+            msg: object,
+        ) -> object:
```
Comment (wch) [inline shiny/session/_session.py:426]:
```suggestion
        ) -> Jsonifiable:
```

File: shiny/session/_session.py [hunk 5]
```diff
@@ -404,6 +417,43 @@ async def _send_response(self, message: ClientMessageOther, value: object) -> No
...
+            """
+            Used to handle request messages from an Output Renderer.
+
+            Typically, this is used when an Output Renderer needs to set an input value.
+            E.g. a @render.data_frame result with a `mode="edit"` parameter will use
+            this to set the value of the data frame when a cell is edited.
+            """
```
Comment (wch) [inline shiny/session/_session.py:433]: Say that the return value will be sent back to the browser.

File: shiny/session/_session.py [hunk 6]
```diff
@@ -404,6 +417,43 @@ async def _send_response(self, message: ClientMessageOther, value: object) -> No
...
+            with session_context(self):
```
Comment (wch) [inline shiny/session/_session.py:434]: Is this necessary?

File: shiny/session/_session.py [hunk 7]
```diff
@@ -404,6 +417,43 @@ async def _send_response(self, message: ClientMessageOther, value: object) -> No
...
+                renderer = output_info.renderer
+                handler = f"_handle_{handler}"
+                if not (
+                    hasattr(renderer, handler) and callable(getattr(renderer, handler))
+                ):
+                    raise RuntimeError(
+                        f"Output Renderer with id `{outputId}` does not have method `{handler}` to handler request message"
+                    )
+                try:
+                    return await getattr(renderer, handler)(msg)
```
Comment (wch) [inline shiny/session/_session.py:450]: Handler methods should manage their own session context, in a way that works with modules.

File: shiny/session/_session.py (hunk 4 above)
Comment (wch) [reply shiny/session/_session.py:426]: Also, move this into `Outputs` class if possible.

File: shiny/session/_session.py [hunk 8]
```diff
@@ -1147,19 +1214,20 @@ async def output_obs():
...
-            del self._suspend_when_hidden[output_name]
+        if output_name in self._outputs:
+            self._outputs[output_name].effect.destroy()
+            del self._outputs[output_name]
 
     def _manage_hidden(self) -> None:
         "Suspends execution of hidden outputs and resumes execution of visible outputs."
-        output_names = list(self._suspend_when_hidden.keys())
-        for name in output_names:
+        for name in self._outputs:
```
Comment (wch) [inline shiny/session/_session.py:1223]:
```suggestion
        for output, name in self._outputs.items():
```

**Comment by schloerke** [comment]
Closing in favor of #1198

**Pull request closed without merging**